File: src/Q33.cpp
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

class Solution {
public:
  bool verifyPostorder(vector<int> &postorder) {
    stack<int> st;
    int root = INT_MAX;
    for (auto it = postorder.rbegin(); it != postorder.rend(); ++it) {
      if (*it > root)
        return false;
      while (!st.empty() && st.top() > *it) {
        root = st.top();
        st.pop();
      }
      st.push(*it);
    }
    return true;
  }
};

File: src/Q33.rs
struct Solution;

impl Solution {
    pub fn verify_postorder(postorder: Vec<i32>) -> bool {
        let mut root = i32::MAX;
        let mut stack = Vec::new();
        for val in postorder.into_iter().rev() {
            if val > root {
                return false;
            }
            while !stack.is_empty() && stack[stack.len() - 1] > val {
                root = stack.pop().unwrap();
            }
            stack.push(val);
        }
        true
    }
}
